### PR TITLE
Exclude development from static html/json render on data request

### PR DESCRIPTION
**Pull request by timneutkens**
As found by @alexkirsz this caused the HTML rendering to happen in development even for data requests of static pages, which is not needed because the incremental cache doesn't apply in development.

Additional updated the launch.json to not say `launch app` given that it might be confused with the `app` directory.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm build && pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| buildDuration | 36.3s | 37s | ⚠️ +757ms |
| buildDurationCached | 15.6s | 10.6s | -4.9s |
| nodeModulesSize | 95.1 MB | 95.1 MB | ✓ |
| nextStartRea..uration (ms) | 281ms | 276ms | -5ms |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| 152-HASH.js gzip | 64.5 kB | 64.5 kB | ✓ |
| 53.HASH.js gzip | 182 B | 182 B | ✓ |
| main-app-HASH.js gzip | 205 B | 205 B | ✓ |
| main-HASH.js gzip | 79.2 kB | 79.2 kB | ✓ |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 146 kB | 146 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| _app-HASH.js gzip | 192 B | 192 B | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 484 B | 484 B | ✓ |
| css-HASH.js gzip | 804 B | 804 B | ✓ |
| dynamic-HASH.js gzip | 2.27 kB | 2.27 kB | ✓ |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 827 B | 827 B | ✓ |
| hooks-HASH.js gzip | 848 B | 848 B | ✓ |
| image-HASH.js gzip | 4.3 kB | 4.3 kB | ✓ |
| index-HASH.js gzip | 252 B | 252 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | ✓ |
| routerDirect..HASH.js gzip | 782 B | 782 B | ✓ |
| script-HASH.js gzip | 857 B | 857 B | ✓ |
| withRouter-HASH.js gzip | 781 B | 781 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.6 kB | 15.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| _buildManifest.js gzip | 482 B | 482 B | ✓ |
| Overall change | 482 B | 482 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| index.html gzip | 489 B | 489 B | ✓ |
| link.html gzip | 504 B | 504 B | ✓ |
| withRouter.html gzip | 485 B | 485 B | ✓ |
| Overall change | 1.48 kB | 1.48 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| edge-ssr.js gzip | 110 kB | 110 kB | ✓ |
| page.js gzip | 98.4 kB | 98.4 kB | ⚠️ +1 B |
| Overall change | 209 kB | 209 kB | ⚠️ +1 B |

</details>

<details>
<summary><strong>Middleware size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js fix/bypass-html-in-dev | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 581 B | 582 B | ⚠️ +1 B |
| middleware-r..fest.js gzip | 145 B | 145 B | ✓ |
| middleware.js gzip | 27 kB | 27 kB | ✓ |
| edge-runtime..pack.js gzip | 1.83 kB | 1.83 kB | ✓ |
| Overall change | 29.6 kB | 29.6 kB | ⚠️ +1 B |

</details>

#### Diffs
<details>
<summary>Diff for <strong>page.js</strong></summary>

```diff
@@ -8618,7 +8618,7 @@ Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime`;
       /***/
     },
 
-    /***/ 6738: /***/ (
+    /***/ 1036: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -8627,10 +8627,10 @@ Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime`;
         __webpack_require__.bind(__webpack_require__, 1210)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.bind(__webpack_require__, 6448)
+        __webpack_require__.bind(__webpack_require__, 2973)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.bind(__webpack_require__, 2973)
+        __webpack_require__.bind(__webpack_require__, 6448)
       );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.bind(__webpack_require__, 9916)
```
</details>
<details>
<summary>Diff for <strong>middleware-b..-manifest.js</strong></summary>

```diff
@@ -9,7 +9,7 @@ self.__BUILD_MANIFEST = {
   rootMainFiles: [
     "static/chunks/webpack-c452a3e31b73f504.js",
     "static/chunks/152-792c0b16c9234019.js",
-    "static/chunks/main-app-41573aefc84f5f2e.js"
+    "static/chunks/main-app-c9a0aaff35eafcb5.js"
   ],
   pages: {
     "/": [
```
</details>
<details>
<summary>Diff for <strong>main-app-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [744],
   {
-    /***/ 2938: /***/ function(
+    /***/ 9915: /***/ function(
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -10,10 +10,10 @@
         __webpack_require__.t.bind(__webpack_require__, 429, 23)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.t.bind(__webpack_require__, 8138, 23)
+        __webpack_require__.t.bind(__webpack_require__, 1161, 23)
       );
       Promise.resolve(/* import() eager */).then(
-        __webpack_require__.t.bind(__webpack_require__, 1161, 23)
+        __webpack_require__.t.bind(__webpack_require__, 8138, 23)
       );
       Promise.resolve(/* import() eager */).then(
         __webpack_require__.t.bind(__webpack_require__, 9379, 23)
@@ -28,7 +28,7 @@
       return __webpack_require__((__webpack_require__.s = moduleId));
     };
     /******/ __webpack_require__.O(0, [152], function() {
-      return __webpack_exec__(7070), __webpack_exec__(2938);
+      return __webpack_exec__(7070), __webpack_exec__(9915);
     });
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
</details>
Commit: 77db6311a501614b6bb1fedfd9db033bc8bc4893

File: packages/next/src/server/render.tsx [hunk 1]
```diff
@@ -1057,7 +1057,8 @@ export async function renderToHTML(
 
   // Avoid rendering page un-necessarily for getServerSideProps data request
   // and getServerSideProps/getStaticProps redirects
-  if ((isDataReq && !isSSG) || (renderOpts as any).isRedirect) {
+  // In development the incrementalCache is not used so we short-circuit bypassing the html rendering.
+  if ((isDataReq && (!isSSG || dev)) || (renderOpts as any).isRedirect) {
```
Comment (ijjk) [inline packages/next/src/server/render.tsx:1061]: Hmm does the benefit of avoiding rendering here in dev offset the potential downside that a rendering error isn't caught?

Comment (timneutkens) [reply packages/next/src/server/render.tsx:1061]: It's specifically needed as Turbopack handles data requests differently in that the team implemented a reverse of the getStaticProps/getServerSideProps tree shaking transform to allow better HMR of getStaticProps/getServerSideProps changes. So you end up with a file that doesn't have the component code, only the code used in getStaticProps.

Comment (ijjk) [reply packages/next/src/server/render.tsx:1061]: Sounds like that will need to be changed soon as we're changing how turbopack is handled and wouldn't work in production.

**Comment by timneutkens** [comment]
Going to close this as we implemented a different solution.

**Pull request closed without merging**